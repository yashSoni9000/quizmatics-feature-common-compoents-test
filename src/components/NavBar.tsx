import { useContext, useState } from "react";

import MenuIcon from "@mui/icons-material/Menu";
import {
    Box,
    IconButton,
    Typography,
    Menu,
    MenuItem,
    Container,
    Avatar,
    Tooltip,
    Switch,
    styled,
} from "@mui/material/";

import { AppContext, AppContextType } from "@src/context/AppProvider";

import QuizLogo from "@images/favicon.png";
import ProfileImg from "@images/profile-img.svg";
import { removeToken } from "@src/utils/token";
import { useNavigate } from "react-router-dom";
import CircularLoader from "./common/CircularLoader";

interface Props {
    displaySideBar: (value: boolean) => void;
}

export default function NavBar({ displaySideBar }: Props) {
    const [anchorElUser, setAnchorElUser] = useState<null | HTMLElement>(null);
    const [showLogoutWindow, setShowLogoutWindow] = useState(false);

    const appContext = useContext<AppContextType>(AppContext)!;
    const navigate = useNavigate();

    const { changeTheme } = appContext;

    const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
        setAnchorElUser(event.currentTarget);
    };
    const handleCloseUserMenu = () => {
        setAnchorElUser(null);
    };

    const settings = [
        {
            label: "Logout",
            onClick: () => {
                removeToken();
                setShowLogoutWindow(true);
                setTimeout(() => {
                    navigate("/login");
                }, 500);
            },
        },
    ];

    return (
        <>
            <Container
                className="flex w-full p-0 h-full items-center max-w-full"
                sx={{
                    backgroundColor: "primary.main",
                    color: "primary.contrastText",
                }}
            >
                <Box className="pl-3 mr-2 hidden lg:flex">
                    <img src={QuizLogo} className="h-10" alt="" />
                </Box>
                <Typography
                    variant="h6"
                    noWrap
                    className="m-4 text-2xl hidden lg:flex"
                >
                    QuizMatics
                </Typography>

                <Box className="flex lg:none">
                    <IconButton
                        size="large"
                        aria-label="menu-appbar"
                        aria-controls="menu-appbar"
                        aria-haspopup="true"
                        onClick={() => {
                            displaySideBar(true);
                        }}
                        className="block lg:hidden"
                        color="inherit"
                    >
                        <MenuIcon />
                    </IconButton>
                </Box>
                <Typography
                    variant="h5"
                    noWrap
                    className="mr-4 font-bold flex lg:hidden"
                >
                    QuizMatics
                </Typography>
                <Box className="ml-auto flex mr-3">
                    <Box className="w-fit h-full mr-2">
                        <AppThemeSwitch
                            checked={localStorage.getItem("theme") !== "light"}
                            onChange={(e) =>
                                changeTheme(
                                    e.currentTarget.checked ? "dark" : "light",
                                )
                            }
                            data-testid={"theme-switch"}
                        />
                    </Box>

                    <Tooltip title="Open settings">
                        <IconButton
                            onClick={handleOpenUserMenu}
                            className="p-0"
                        >
                            <Avatar alt="Remy Sharp" src={ProfileImg} />
                        </IconButton>
                    </Tooltip>
                    <Menu
                        className="mt-12"
                        id="menu-appbar"
                        anchorEl={anchorElUser}
                        anchorOrigin={{
                            vertical: "top",
                            horizontal: "right",
                        }}
                        keepMounted
                        transformOrigin={{
                            vertical: "top",
                            horizontal: "right",
                        }}
                        open={Boolean(anchorElUser)}
                        onClose={handleCloseUserMenu}
                    >
                        {settings.map((setting) => (
                            <MenuItem
                                key={setting.label}
                                onClick={() => {
                                    handleCloseUserMenu();
                                    setting.onClick();
                                }}
                            >
                                <Typography textAlign="center">
                                    {setting.label}
                                </Typography>
                            </MenuItem>
                        ))}
                    </Menu>
                </Box>
            </Container>
            <Box
                className={`${
                    showLogoutWindow ? "flex flex-col " : "hidden"
                } absolute top-0 left-0 w-screen h-screen justify-center  items-center backdrop-blur-sm gap-2 z-30`}
                data-testid={"logoutScreen"}
            >
                <CircularLoader />
                <Typography
                    className="text-base z-40 opacity-100"
                    sx={{ color: "primary.main" }}
                >
                    Logging Out..
                </Typography>
            </Box>
        </>
    );
}

const AppThemeSwitch = styled(Switch)(({ theme }) => ({
    width: 48,
    height: 48,
    padding: 8,
    "& .MuiSwitch-switchBase": {
        margin: 0,
        padding: 7,
        transform: "translateX(0px)",
        "&.Mui-checked": {
            color: "#fff",
            transform: "translateX(0px)",
            "& .MuiSwitch-thumb:before": {
                backgroundImage: `url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" height="20" width="20" viewBox="0 0 20 20"><path fill="${encodeURIComponent(
                    "#fff",
                )}" d="M4.2 2.5l-.7 1.8-1.8.7 1.8.7.7 1.8.6-1.8L6.7 5l-1.9-.7-.6-1.8zm15 8.3a6.7 6.7 0 11-6.6-6.6 5.8 5.8 0 006.6 6.6z"/></svg>')`,
            },
            "& + .MuiSwitch-track": {
                opacity: 1,
                backgroundColor: "#fff",
            },
        },
    },
    "& .MuiSwitch-thumb": {
        backgroundColor: theme.palette.mode === "dark" ? "#1976d2" : "#161C34",

        width: 32,
        height: 32,
        "&:before": {
            content: "''",
            position: "absolute",
            width: "100%",
            height: "100%",
            left: 0,
            top: 0,
            backgroundRepeat: "no-repeat",
            backgroundPosition: "center",
            backgroundImage: `url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" height="20" width="20" viewBox="0 0 20 20"><path fill="${encodeURIComponent(
                "#fff",
            )}" d="M9.305 1.667V3.75h1.389V1.667h-1.39zm-4.707 1.95l-.982.982L5.09 6.072l.982-.982-1.473-1.473zm10.802 0L13.927 5.09l.982.982 1.473-1.473-.982-.982zM10 5.139a4.872 4.872 0 00-4.862 4.86A4.872 4.872 0 0010 14.862 4.872 4.872 0 0014.86 10 4.872 4.872 0 0010 5.139zm0 1.389A3.462 3.462 0 0113.471 10a3.462 3.462 0 01-3.473 3.472A3.462 3.462 0 016.527 10 3.462 3.462 0 0110 6.528zM1.665 9.305v1.39h2.083v-1.39H1.666zm14.583 0v1.39h2.084v-1.39h-2.084zM5.09 13.928L3.616 15.4l.982.982 1.473-1.473-.982-.982zm9.82 0l-.982.982 1.473 1.473.982-.982-1.473-1.473zM9.305 16.25v2.083h1.389V16.25h-1.39z"/></svg>')`,
        },
    },
    "& .MuiSwitch-track": {
        opacity: 1,
        height: 32,
        backgroundColor: theme.palette.mode === "dark" ? "#fff" : "#fff",
        borderRadius: 32,
        transform: "translate(-1px,-1px)",
    },
}));
