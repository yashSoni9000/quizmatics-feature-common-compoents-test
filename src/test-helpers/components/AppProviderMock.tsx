import { AppContext, AppContextType } from "@context/AppProvider";
import { lightTheme } from "@src/themes/theme";
import { Role } from "@src/ts/enums";
import { PropsWithChildren } from "react";

type Props = {
    value?: Partial<AppContextType>;
};

const defaultValues: AppContextType = {
    userDetails: {
        id: 1,
        email: "",
        role: Role.Host,
    },
    currentTheme: lightTheme,
    updateUserDetails: jest.fn(),
    changeTheme: jest.fn(),
    refreshUserDetails: jest.fn(),
};

const AppProviderMock = ({
    children,
    value = {},
}: PropsWithChildren<Props>) => {
    return (
        <AppContext.Provider value={{ ...defaultValues, ...value }}>
            {children}
        </AppContext.Provider>
    );
};
export default AppProviderMock;
