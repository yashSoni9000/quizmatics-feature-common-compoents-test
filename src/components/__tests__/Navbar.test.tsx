import { fireEvent, render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import "@testing-library/jest-dom";

import NavBar from "../NavBar";
import AppProviderMock from "@src/test-helpers/components/AppProviderMock";

const mockedDisplaySidebar = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.useFakeTimers();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUsedNavigate,
}));

describe("Navbar Test", () => {
    it("snapshot test", () => {
        const { asFragment } = render(
            <AppProviderMock>
                <NavBar displaySideBar={mockedDisplaySidebar} />
            </AppProviderMock>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("Profile menu is opened when clicked on it", () => {
        render(
            <AppProviderMock>
                <NavBar displaySideBar={mockedDisplaySidebar} />,
            </AppProviderMock>,
        );
        const profileButton = screen.getByLabelText("Open settings");
        const logoutButton = screen.getByText("Logout");

        fireEvent.click(profileButton);

        expect(logoutButton).toBeVisible();
    });

    it("Logout button logs out the user", async () => {
        render(
            <AppProviderMock>
                <NavBar displaySideBar={mockedDisplaySidebar} />
            </AppProviderMock>,
        );

        const profileButton = screen.getByLabelText("Open settings");
        fireEvent.click(profileButton);

        const logoutButton = screen.getByText("Logout");
        const logoutScreen = screen.getByTestId("logoutScreen");

        expect(logoutScreen).toHaveClass("hidden");
        fireEvent.click(logoutButton);
        expect(logoutScreen).toHaveClass("flex");

        act(() => {
            jest.runAllTimers();
        });
        await expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
    });

    it("Clicking on theme button change theme", async () => {
        const mockedChangeTheme = jest.fn();

        render(
            <AppProviderMock value={{ changeTheme: mockedChangeTheme }}>
                <NavBar displaySideBar={mockedDisplaySidebar} />
            </AppProviderMock>,
        );

        const themeSwitchElement = screen.getByTestId("theme-switch");
        const themeSwitch = themeSwitchElement.querySelector("input");

        if (!themeSwitch) throw new Error();

        fireEvent.change(themeSwitch, { target: { checked: true } });
        fireEvent.click(themeSwitch);

        expect(mockedChangeTheme).toHaveBeenNthCalledWith(1, "light");

        fireEvent.change(themeSwitch, { target: { checked: false } });
        fireEvent.click(themeSwitch);

        expect(mockedChangeTheme).toHaveBeenNthCalledWith(2, "dark");
    });
});
