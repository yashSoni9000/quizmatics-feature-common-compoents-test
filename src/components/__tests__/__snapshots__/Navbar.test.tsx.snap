// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Test snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg flex w-full p-0 h-full items-center max-w-full css-1uuj62e-MuiContainer-root"
  >
    <div
      class="pl-3 mr-2 hidden lg:flex MuiBox-root css-0"
    >
      <img
        alt=""
        class="h-10"
        src="favicon.png"
      />
    </div>
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap m-4 text-2xl hidden lg:flex css-8u39c-MuiTypography-root"
    >
      QuizMatics
    </h6>
    <div
      class="flex lg:none MuiBox-root css-0"
    >
      <button
        aria-controls="menu-appbar"
        aria-haspopup="true"
        aria-label="menu-appbar"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge block lg:hidden css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap mr-4 font-bold flex lg:hidden css-1961ak2-MuiTypography-root"
    >
      QuizMatics
    </h5>
    <div
      class="ml-auto flex mr-3 MuiBox-root css-0"
    >
      <div
        class="w-fit h-full mr-2 MuiBox-root css-0"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-1ibuyee-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            data-testid="theme-switch"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
      </div>
      <button
        aria-label="Open settings"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium p-0 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
        >
          <img
            alt="Remy Sharp"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="profile-img.svg"
          />
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="hidden absolute top-0 left-0 w-screen h-screen justify-center  items-center backdrop-blur-sm gap-2 z-30 MuiBox-root css-0"
    data-testid="logoutScreen"
  >
    <div
      class="flex justify-center items-center  MuiBox-root css-0"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 text-base z-40 opacity-100 css-vkc4v8-MuiTypography-root"
    >
      Logging Out..
    </p>
  </div>
</DocumentFragment>
`;
